kayenta-prometheus/src/main/java/com/netflix/kayenta/prometheus/metrics/PrometheusMetricDescriptorsCache.java

```diff
@@ -0,0 +1,131 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.kayenta.prometheus.metrics;
+
+import com.netflix.kayenta.prometheus.model.PrometheusMetricDescriptor;
+import com.netflix.kayenta.prometheus.model.PrometheusMetricDescriptorsResponse;
+import com.netflix.kayenta.prometheus.security.PrometheusNamedAccountCredentials;
+import com.netflix.kayenta.prometheus.service.PrometheusRemoteService;
+import com.netflix.kayenta.security.AccountCredentials;
+import com.netflix.kayenta.security.AccountCredentialsRepository;
+import lombok.Value;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.util.CollectionUtils;
+import org.springframework.util.StringUtils;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Slf4j
+public class PrometheusMetricDescriptorsCache {
+
+    private volatile Map<String, List<PrometheusMetricDescriptor>> cache = Collections.emptyMap();
+
+    private final AccountCredentialsRepository accountCredentialsRepository;
+
+    public PrometheusMetricDescriptorsCache(AccountCredentialsRepository accountCredentialsRepository) {
+        this.accountCredentialsRepository = accountCredentialsRepository;
+    }
+
+    public List<Map> getMetadata(String metricsAccountName, String filter) {
+        List<PrometheusMetricDescriptor> accountSpecificMetricDescriptorsCache =
+                this.cache.get(metricsAccountName);
+
+        if (CollectionUtils.isEmpty(accountSpecificMetricDescriptorsCache)) {
+            return Collections.emptyList();
+        }
+
+        if (StringUtils.isEmpty(filter)) {
+            return accountSpecificMetricDescriptorsCache.stream()
+                    .map(metricDescriptor -> metricDescriptor.getMap())
+                    .collect(Collectors.toList());
+        } else {
+            String lowerCaseFilter = filter.toLowerCase();
+
+            return accountSpecificMetricDescriptorsCache.stream()
+                    .filter(
+                            metricDescriptor ->
+                                    metricDescriptor.getName().toLowerCase().contains(lowerCaseFilter))
+                    .map(metricDescriptor -> metricDescriptor.getMap())
+                    .collect(Collectors.toList());
+        }
+    }
+
+    @Scheduled(fixedDelayString = "#{@prometheusConfigurationProperties.metadataCachingIntervalMS}")
+    public void updateMetricDescriptorsCache() {
+        Set<AccountCredentials> accountCredentialsSet =
+                accountCredentialsRepository.getAllOf(AccountCredentials.Type.METRICS_STORE);
+
```

Comment:
replaced existing code with the streams api without any changes in logic. this way it looks more simple